lib/protobuf/json/decode.ex

```diff
@@ -0,0 +1,195 @@
+defmodule Protobuf.JSON.Decode do
+  @moduledoc false
+
+  alias Protobuf.JSON.Utils
+
+  @compile {:inline,
+            field_value: 2,
+            decode_integer: 1,
+            decode_float: 1,
+            decode_bytes: 1,
+            decode_enum: 2,
+            decode_key: 3,
+            parse_key: 2,
+            maybe_repeat: 3}
+
+  # TODO: optimize order by popularity
+  @int_types ~w(int32 sint32 sfixed32 fixed32 uint32 int64 sint64 sfixed64 fixed64 uint64)a
+  @float_types [:float, :double]
+
+  def from_json_data(data, module) when is_map(data) do
+    message_props = Utils.message_props(module)
+    regular = decode_regular_fields(data, message_props)
+    oneofs = decode_oneof_fields(data, message_props)
+
+    module.__default_struct__()
+    |> struct(regular)
+    |> struct(oneofs)
+  end
+
+  def from_json_data(data, _module), do: throw({:bad_message, data})
+
+  defp decode_regular_fields(data, %{field_props: field_props}) do
+    for {_field_num, %{oneof: nil} = prop} <- field_props,
+        value = field_value(prop, data) do
+      {prop.name_atom, decode_value(prop, value)}
+    end
+  end
+
+  defp decode_oneof_fields(data, %{field_props: field_props, oneof: oneofs}) do
+    for {oneof, index} <- oneofs,
+        {_field_num, %{oneof: ^index} = prop} <- field_props,
+        value = field_value(prop, data),
+        reduce: %{} do
```

Comment:
This is Elixir 1.8+ only. I can do it with a regular reduce if we need to support 1.7.